Mas i370 patch d

martinsumner opened the pull request:
https://github.com/martinsumner/leveled/issues/370

Previous attempts to manage memory usage in large stores has focused on the Block Index Cache, which is about 2 bytes per key (for directly fetchable keys i.e. not index entries) in the leveled_sst files.  These caches grow through use, but were never expunged - and perhaps files with mainly dead data would falsely retain these caches without performance benefits.

The balance is difficult though, between clearing these caches and the non-functional value in  retaining them (especially wrt to increasing the speed of AAE-related queries).

Running a 36 hour test with relatively small values, to build up a key store.  

Three things are noted:

- Memory fragmentation across the nodes on the cluster is between 52% and 56%.
- The memory footprint of leveled_sst processes is still large, even when the Block Index Cache has been expunged (following journal compaction).
- The memory footprint of leveled _sst files (as measured by `process_info(P, memory)`) is commonly about 8 times larger than the footprint of process state (as measured by `erts_debug:flat_size(sys:get_state(P))`) even after forced garbage collection on the PID.  It is some times 25 times larger.  This is the case even when binary memory owned by the process is empty (as measured by `process_info(P, binary)`). **EDIT ... IGNORE THIS! `erts_debug:flat_size/1` is measured in words not bytes!  One word equals 8 bytes on a 64-bit system**

This PR is a combination of three refactoring exercises intended to try and improve this situation:

- https://github.com/martinsumner/leveled/commit/d393c99856ba79dff2da0b83157f43eb4e9cb6e2 which refactors the penciller memory.  Partly for readability (to stop LoopState being passed to internal functions), but also to try and do memory efficient operations where possible (e.g. `[Addition|ExistingList]` to extend lists).
- https://github.com/martinsumner/leveled/commit/415e682c6d8d2b9c9354a05f487abd2984e4dc44 which changes the index kept within a leveled_sst file (of 1 in 128 keys).  This now attempts to find common elements in this index, (e.g. where all entries in the SST file have the same Type/Bucket) and then only index the deltas.
- https://github.com/martinsumner/leveled/commit/56c5e2565ddde15edebd7c0097cb57882b91d3ed which fixes a bug in the Block Index Cache which meant that when a BIC is cleared, if memory of the HMD date for the file is retained, the cache would never be rebuilt. The refactoring also reduces unnecessary changes to the BIC as the BIC is populated, and an unnecessary fold over the BIC to rediscover the HMD.

It is hoped that tidying up these elements will make it easier to investigate further memory overheads (and perhaps in the case of the leveled_sst index changes, directly improve memory footprint).

martinsumner committed:
Refactor penciller memory

In high-volume tests on large key-count clusters, so significant variation in the P0031 time has been seen:

TimeBucket	PatchA
a.0ms_to_1ms	18554
b.1ms_to_2ms	51778
c.2ms_to_3ms	696
d.3ms_to_5ms	220
e.5ms_to_8ms	59
f.8ms_to_13ms	40
g.13ms_to_21ms	364
h.21ms_to_34ms	277
i.34ms_to_55ms	34
j.55ms_to_89ms	17
k.89ms_to_144ms	21
l.144ms_to_233ms	31
m.233ms_to_377ms	45
n.377ms_to_610ms	52
o.610ms_to_987ms	59
p.987ms_to_1597ms	55
q.1597ms_to_2684ms	54
r.2684ms_to_4281ms	29
s.4281ms_to_6965ms	7
t.6295ms_to_11246ms	1

It is unclear why this varies so much.  The time to add to the cache appears to be minimal (but perhaps there is an issue with timing points in the code), whereas the time to add to the index is much more significant and variable.  There is also variable time when the memory is rolled (although the actual activity here appears to be minimal.

The refactoring here is two-fold:

- tidy and simplify by keeping LoopState managed within handle_call, and add more helpful dialyzer specs;

- change the update to the index to be a simple extension of a list, rather than any conversion.

This alternative version of the pmem index in unit test is orders of magnitude faster to add - and is the same order of magnitude to check.  Anticipation is that it may be more efficient in terms of memory changes.

martinsumner committed:
Compress SST index

Reduces the size of the leveled_sst index with two changes:

1 - Where there is a common prefix of tuple elements (e.g. Bucket) across the whole leveled_sst file - only the non-common part is indexed, and a function is used to compare.

2 - There is less "indexing" of the index i.e. only 1 in 16 keys are passed into the gb_trees part instead of 1 in 4

martinsumner committed:
Immediate hibernate

Reasons for delay in hibernate were not clear.

Straight after creation the process will not be in receipt of messages (must wait for the manifest to be updated), so better to hibernate now.  This also means the log PC023 provides more accurate information.

martinsumner committed:
Refactor BIC

This patch avoids the following:

- repeated replacement of the same element in the BIC (via get_kvrange), by checking presence via GET before sing SET

- Stops re-reading of all elements to discover high modified date

Also there appears to have been a bug where a missing HMD for the file is required to add to the cache.  However, now the cache may be erased without erasing the HMD.  This means that the cache can never be rebuilt

martinsumner committed:
Use correct size in test results

erts_debug:flat_size/1 returns size in words (i.e. 8 bytes on 64-bit CPU) not bytes

martinsumner committed:
Don't change summary record

As it is persisted as part of the file write, any change to the summary record cannot be rolled back

martinsumner committed:
Clerk to prompt L0 write

Simplifies the logic if the clerk request work for the penciller prompts L0 writes as well as Manifest changes.

The advantage now is that if the penciller memory is full, and PUT load stops, the clerk should still be able to prompt persistence.  the penciller can therefore make use of dead time this way

martinsumner committed:
Add push on journal compact

If there has been a backlog, followed by a quiet period - there may be a large ledger cache left unpushed.  Journal compaction events are about once per hour, so the performance overhead of a false push should be minimal, with the advantage of clearing any backlog before load starts again.

This is only relevant to riak users with very off/full batch type workloads.

martinsumner committed:
Extend tests

To more consistently trigger all overload scenarios

martinsumner committed:
Fix range keys smaller than prefix

Can't make end key an empty binary  in this case, as it may be bigger than any keys within the range, but will appear to be smaller.

Unit tests and ct tests added to expose the potential issue

martinsumner committed:
Tidy-up

- Remove penciller logs which are no longer called
- Get pclerk to only wait MIN_TIMEOUT after doing work, in case there is a backlog
- Remove update_levelzero_cache function as it is unique to handle_call of push_mem, and simple enough to be inline
- Alight testutil slow offer with standard slow offer used

martinsumner committed:
Tidy-up

Remove pre-otp20 references.

Reinstate the check that the starting pid is still active, this was added to tidy up shutdown.

Resolve failure to run on otp20 due to `-if` sttaement

martinsumner committed:
Tidy up

Using null rather then {null, Key} is potentially clearer as it is not a concern what they Key is in this case, and removes a comparison step from the leveled_codec:endkey_passed/2 function.

There were issues with coverage in eunit tests as the leveled_pclerk shut down.  This prompted a general tidy of leveled_pclerk (remove passing of LoopState into internal functions, and add dialyzer specs.

martinsumner committed:
Remove R16 relic

martinsumner committed:
Further testing another issue

The StartKey must always be less than or equal to the prefix when the first N characters are stripped,  but this is not true of the EndKey (for the query) which does not have to be between the FirstKey and the LastKey.

If the EndKey query does not match it must be greater than the Prefix (as otherwise it would not have been greater than the FirstKey - so set to null.

martinsumner commented:
Extend Patch C to ensure that both an excess bookie ledger cache, and an over-size penciller in-memory cache will be persisted to disk even where no traffic is present

martinsumner committed:
Fix unit test

Unit test had a typo - and result interpretation had a misunderstanding.

martinsumner committed:
Code and spec tidy

Also look to the cover the situation when the FirstKey is the same as the Prefix with tests.

This is, in theory, not an issue as it is the EndKey for each sublist which is indexed in leveled_tree.  However, guard against it mapping to null here, just in case there are dangers lurking (note that tests will still pass without `M > N` guard in place.

martinsumner committed:
Hibernate on BIC complete

There are three situations when the BIC becomes complete:

- In a file created as part of a merge the BIS is learned in the merge
- After startup, files below L1 learn the block cache through reads that happen to read the block, eventually the while cache will be read, unless...
- Either before/after the cache is complete, it can get whiped by a timeout after a get_sqn request (e.g. as prompted by a journal compaction) ... it will then be re-filled of the back of get/get-range requests.

In all these situations we want to hibernate after the BIC is fill - to reflect the fact that the LoopState should now be relatively stable, so it is a good point to GC and rationalise location of data.

Previously on the the first base was covered.  Now all three are covered through the bic_complete message.

martinsumner commented:
This compares different configurations on a high-key-volume load test (one using small objects and lots of PUTs to try and rapidly increase the volume of keys).

Memory -

<img width="1604" alt="Riak310PatchD_Memory" src="https://user-images.githubusercontent.com/1628897/202717098-94d9b2d3-6457-41e2-a848-b163509c5798.png">

Throughput -

<img width="1602" alt="Riak310PatchD_Throughput" src="https://user-images.githubusercontent.com/1628897/202717135-a6712998-bab9-4893-88ff-2e3554f5e14c.png">

It seems that we get improvements in memory efficiency almost for free in terms of throughput.

The Profile A refers to setting eheap SBCT to 128KB (default 512KB): erlang.eheap_memory.sbct = 128

martinsumner commented:
The memory is as reported by the OS, averaged across the nodes in the cluster.  The test duration was 36 hours

yorkshire-steve commented on src/leveled_sst.erl:
```diff
@@ -1725,20 +1764,100 @@ hmac(Int) when is_integer(Int) ->
+        case Term of
+            T when byte_size(T) =< N ->
+                null;
+            <<QueryKey:N/binary, Suffix/binary>> ->
+                case QueryKey of
```
Should this be QueryTerm as opposed to QueryKey for clarity?

martinsumner replied:
I think the interim assignment to anything is actually unnecessary.  We actually only have two cases, where it maps to <<Prefix:N/binary, Suffix/binary>> ... or where the result should be null.  This is the same in the key_prefix_filter/2 and the term_prefix_filter/2.

martinsumner replied:
https://github.com/martinsumner/leveled/pull/383/commits/76ed44dea489d1fd009d9d6123d92199deb7a72a

yorkshire-steve replied:
I think this change allows `Suffix` to be null (when Key == Prefix), which wasn't allowed previously due to `K when byte_size(K) =< N`
I don't think this will cause a problem though as the return Suffix would be null which is what was being returned previously
Just wanted to point it out in case you knew of any unintended side effect here

martinsumner replied:
In this case the Suffix would be <<>> and not null I think.  However, this is a good point.  Let me think

yorkshire-steve commented on src/leveled_sst.erl:
```diff
@@ -1725,20 +1764,100 @@ hmac(Int) when is_integer(Int) ->
+get_filterfun(
+        {Tag, Bucket, {Field, FT}, FK}, {Tag, Bucket, {Field, LT}, LK})
+            when is_binary(FT), is_binary(FK), is_binary(LT), is_binary(LK) ->
+    case {binary:longest_common_prefix([FT, LT]), byte_size(FT)} of
+        {N, M} when N > 0, M > N ->
```
I keep looking at this and wondering if there could ever be a situation where FT == LT

martinsumner replied:
If there are a lot of keys with exactly the same term on an index, then yes.  That needs a unit test

yorkshire-steve replied:
In that case when FT==LT, won't M==N? Then the test M > N is false and this won't use the more optimised FilterFun

martinsumner replied:
That is correct.  So it works functionally (as it switches to using {Term, Key} on the index), but is not optimised.  I've added a unit test to show that it at least works.

I don't think changing M > N to M >= N will work, so we would need another clause to optimise.

martinsumner replied:
See https://github.com/martinsumner/leveled/pull/383/commits/9db486671cf5b2576ea33fb5d1dbd639c499d7aa

martinsumner committed:
Test all index keys have same term

This works functionally, but is not optimised (the term is replicated in the index)

martinsumner committed:
Summaries with same index term

If the summary index all have the same index term - only the object keys need to be indexes

martinsumner commented:
Tests run to include the latest version of PatchD, and also an additional control on OTP 25 (without leveled fixes):

<img width="1599" alt="Riak310PatchD_Memory2" src="https://user-images.githubusercontent.com/1628897/203300109-7133d670-57fc-4eb9-ad1b-c7e7c6bda17c.png">

The positive news is increased memory efficiency in OTP 25.  However, this doesn't deliver the same level of improvements as the patch

martinsumner committed:
Simplify case statements

We either match the pattern of <<Prefix:N, Suffix>> or the answer should be null

yorkshire-steve commented on src/leveled_sst.erl:
```diff
@@ -1725,20 +1764,107 @@ hmac(Int) when is_integer(Int) ->
+get_filterfun(
+        {Tag, Bucket, FK, null}, {Tag, Bucket, LK, null})
+            when is_binary(FK), is_binary(LK), FK < LK ->
+    case {binary:longest_common_prefix([FK, LK]), byte_size(FK)} of
+        {N, M} when N > 0, M > N ->
+            <<Prefix:N/binary, _Rest/binary>> = FK,
+            key_prefix_filter(N, Prefix);
+        _ ->
+            fun key_filter/1
```
I can't quite remember the rules on ordering, but is it possible to have FK="0000" and LK="0000999999"? or does "0000" come after? if it's the first, then I don't think this would result in an optimised FilterFun. In the 2nd case, this may need a test case to ensure expected behaviour (there may be one already I've not spotted)

martinsumner replied:
So <<"0000">> is smaller than <<"0000">> with anything appended so <<"0000999999">> is > than <<"0000">>.

I think the fact that not every case is optimised is fine, it isn't going to have a material impact if the occasional SST file is using more memory than is strictly required.  So I would rather have simpler code than have full optimisation.

The unit test `key_matchesprefix_test/0` In both the index case and object case creates a First Key that is equal in size to the longest_common_prefix.  But I perhaps added this test as I assumed it would be optimised, when no in fact it just becomes a `key_filter/1` test.

Need to clear my head of other things, and just think this through carefully again.

martinsumner replied:
I'm sure M == N is ok.  The guard M > N was introduced before the Prefix check was added to the filter functions (we were just truncating for size).  The weirdness that might have been caused in these cases is now handled by setting to null whenever the first N don't match the Prefix.

I've updated in https://github.com/martinsumner/leveled/pull/383/commits/0fe735fa823daf9f92e2d468783fbec64483bd49 and all tests pass.

martinsumner committed:
OK for M == N

If M = N for the first key, it will have a suffix of <<>>.  This will match (as expected) a query Start Key of the sam size, and be smaller than any query Start Key that has the same prefix.

If the query Start Key does not match the prefix - it will be null - as it must be smaller than the Prefix (as other wise the query Start Key would be bigger than the Last Key).

The constraint of M > N was introduced before the *_prefix_filter functions were checking the prefix, to avoid issues.  Now the prefix is being checked, then M == N is ok.

martinsumner committed:
Simplify

Correct the test to use a binary field in the range.

To avoid further issue, only apply filter when everything is a binary() type.

martinsumner committed:
Add test for head_only mode

When leveled is used as a tictacaae key store (in parallel mode), the keys will be head_only entries.  Double check they are handled as expected like object keys

martinsumner committed:
Revert previous change - must support typed buckets

Add assertion to confirm worthwhile optimisation

martinsumner commented:
Further test.  This starts from a cluster pre-populated with 360M objects, and then aggressively inserts more over 24 hours (with load shifting to be more read-based in the last hour).  

Memory footprint:

<img width="1601" alt="Riak310PatchD_Memory3" src="https://user-images.githubusercontent.com/1628897/204860193-f5fb539d-5304-4e0d-8b78-fc3ad0a074e7.png">

Throughput achieved:

<img width="1606" alt="Riak310PatchD_Throughput3" src="https://user-images.githubusercontent.com/1628897/204860327-1b55c37a-1d1d-4ad3-b1f4-fb2057ce6f5a.png">

The PatchD performs well in terms of both throughput and memory optimisation.  Adding the alternative configuration profile (sbct = 128KB ), reduces throughput at the start of the tense - but consistently delivers a lower memory footprint.

A longer test needs to be run from the same base, to understand the longer-term behaviour when load goes primarily GETs not PUTs.
